src/encoder/text.rs

```diff
@@ -224,26 +218,49 @@ fn label_pairs_to_text(
 
 /// `escape_string` replaces `\` by `\\`, new line character by `\n`, and `"` by `\"` if
 /// `include_double_quote` is true.
-fn escape_string(v: &str, include_double_quote: bool) -> String {
-    let mut escaped = String::with_capacity(v.len() * 2);
+///
+/// Implementation adapted from
+/// https://lise-henry.github.io/articles/optimising_strings.html
```

Comment:
I somehow have the hope that regex `replace_all` performance improved over that 2016 page, as it changed in the meanwhile to work on Cow. Did you try benchmarking both here?

Reply:
I hoped as well, but the diff below degrades performance extensively:

```bash
$ cargo benchcmp optimized optimized_replace_all 
 name                                 optimized ns/iter  optimized_replace_all ns/iter  diff ns/iter   diff %  speedup 
 bench_text_encoder_with_escaping     105,675,740        253,291,262                     147,615,522  139.69%   x 0.42 
 bench_text_encoder_without_escaping  73,070,167         90,540,884                       17,470,717   23.91%   x 0.81
```

Am I missing something?

```diff
diff --git a/src/encoder/text.rs b/src/encoder/text.rs
index 97a5ca5..f8179aa 100644
--- a/src/encoder/text.rs
+++ b/src/encoder/text.rs
@@ -1,6 +1,6 @@
 // Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
 
-use regex::{Match, Regex};
+use regex::{Captures, Match, Regex};
 use std::borrow::Cow;
 use std::io::Write;
 
@@ -218,9 +218,6 @@ fn label_pairs_to_text(
 
 /// `escape_string` replaces `\` by `\\`, new line character by `\n`, and `"` by `\"` if
 /// `include_double_quote` is true.
-///
-/// Implementation adapted from
-/// https://lise-henry.github.io/articles/optimising_strings.html
 fn escape_string(v: &str, include_double_quote: bool) -> Cow<'_, str> {
     // Regex compilation is expensive. Use `lazy_static` to compile the regexes
     // once per process lifetime and not once per function invocation.
@@ -229,37 +226,18 @@ fn escape_string(v: &str, include_double_quote: bool) -> Cow<'_, str> {
         static ref QUOTED_ESCAPER: Regex = Regex::new("(\\\\|\n|\")").expect("Regex to be valid.");
     }
 
-    let first_occurence = if include_double_quote {
-        QUOTED_ESCAPER.find(v)
-    } else {
-        ESCAPER.find(v)
-    }
-    .as_ref()
-    .map(Match::start);
-
-    if let Some(first) = first_occurence {
-        let mut escaped = String::with_capacity(v.len() * 2);
-        escaped.push_str(&v[0..first]);
-        let remainder = v[first..].chars();
-
-        for c in remainder {
-            match c {
-                '\\' | '\n' => {
-                    escaped.extend(c.escape_default());
-                }
-                '"' if include_double_quote => {
-                    escaped.extend(c.escape_default());
-                }
-                _ => {
-                    escaped.push(c);
-                }
-            }
+    let replacer = |cap: &Captures| {
+        match &cap[0] {
+            c@ "\\" | c@"\n" => c.escape_default().to_string(),
+            c@"\"" if include_double_quote => c.escape_default().to_string(),
+            c @ _ => panic!("Got unexpected capture {:?}.", c),
         }
-        escaped.into()
+    };
+
+    if include_double_quote {
+        QUOTED_ESCAPER.replace_all(v, replacer)
     } else {
-        // The input string does not contain any characters that would need to
-        // be escaped. Return it as it is.
-        v.into()
+        ESCAPER.replace_all(v, replacer)
     }
 }
```

Reply:
I tried some further changes to the replacer, but I never got any better than 0.58x than your PR.

Reply:
Out of curiosity, what were your changes @lucab?

Reply:
@mxinden dropping the group/capturing in regex and plain `str` replacements arms (like `"\\" =>  "\\\\"`). Sorry I've already thrown away that branch.